internal/datastore/proxy/cachedcheckrev.go

```diff
@@ -0,0 +1,44 @@
+package proxy
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+)
+
+// newCachedCheckRevision wraps a datastore with a cache that will avoid checking the revision
+// if the last checked revision is at least as fresh as the one specified.
+func newCachedCheckRevision(ds datastore.ReadOnlyDatastore) datastore.ReadOnlyDatastore {
```

Comment:
I understand the desire to limit the use to the most strict subset of the functionality, but I'd add, as we discussed, that this can work with a primary too, but we are not going to reduce the surface of the change in this PR.

Reply:
Yep